### DISCUSSION TITLE
How to handle relationship conflicts on join table with matching model

### DISCUSSION THREAD

**Discussion by stefanv** [sqlalchemy/sqlalchemy#7967] [Usage Questions]
I apologize in advance for having to take up your time with this question, but I've now spent many hours reading the docs and scouring the internet, and cannot find a solution.

I have a model Role, a model ACL, and then do a many-to-many join between them. In that process, I generate a model RoleACL to view/edit the join table.

Since upgrading to SQLAlchemy 1.4 I see:

```
SAWarning: relationship 'RoleACL.role' will copy column roles.id to column role_acls.role_id, which conflicts with relationship(s): 'Role.acls' (copies roles.id to role_acls.role_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="acls"' to the 'RoleACL.role' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)
```

The docs talk about this exact issue, but I'm not at all sure how to address it correctly.

You can see the models here: https://github.com/cesium-ml/baselayer/blob/71223a4abe83d5231cf334481267fe6c629be20b/app/models.py#L1178

And the code which generates the join table/model here:

https://github.com/cesium-ml/baselayer/blob/71223a4abe83d5231cf334481267fe6c629be20b/app/models.py#L1079

I presume having a model for a join table is fairly common, so hopefully an answer to my question will also be helpful to others.

**Comment by zzzeek** [comment]
> relationship 'RoleACL.role' will copy column roles.id to column role_acls.role_id, which conflicts with relationship(s): 'Role.acls' (copies roles.id to role_acls.role_id). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="acls"' to the 'RoleACL.role' relationship. (Background on this error at: https://sqlalche.me/e/14/qzyx)

just from reading the names, it sounds like you have a relationship RoleACL.role and Role.acls, which are a many-to-one / one-to-many pair, however you don't have back_populates set up between them.  This means you might assign Role(id=1) to a certain RoleACL(id=5).role, and then you might append that same RoleACL(id=5) to the "acls" collection of Role(id=2).   When the unit of work flushes, it's essentially random if RoleACL(id=5) will be assigned role_id=1 or role_id=2.   This is what the warning is indicating can happen.

ok right, you have Role.acls which is set up to be ready to write data into role_acls as secondary, and the code at https://github.com/cesium-ml/baselayer/blob/71223a4abe83d5231cf334481267fe6c629be20b/app/models.py#L1145 is generating RoleACL.acls which is also ready to write values into this table, so the ORM sees these two relationships as capable of persisting contradictory information in a single row.   in particular, many-to-many overlapping with association object pattern can lead to this happening if your runtime code is writing to both relationships.  This is discussed in the note at https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object, the pink box with the dragon in it at the bottom.

short answer, put viewonly=True on the relationship that you arent using for persistence.

**Reply by stefanv** [reply]
Thank you for looking at this @zzzeek!

The way we use the join table is to link two instances together. For example, we add users to groups by using:

```
GroupUser(user_id=user.id, group_id=group.id)
```

But for the rest we only work with `Group` or `User`.  So, given that we never *create* Groups or Users this way, are you saying that the view_only relationships will be on the joined model?

I.e., we will have:

```
Role.acls -> back-propagate to ACL.roles
ACL.roles -> back-propagate to Role.acls
```

And then we want RoleACL be able to insert entries in the join table, but do no other back-propagation.

**Reply by zzzeek** [reply]
if you are creating RoleACL(role_id=x, acl_id=y) and that's the only way you INSERT into this table, then all the relationships in question could be viewonly, because you aren't making use of any of the persistence features of relationship.

But, you seem to have passive_deletes set up on Role.acls, which implies that you *are* using persistence features of this relationship.  if you set the relationships to viewonly, you could use ON DELETE CASCADE on the foreign key constraints on roleacl instead, that would handle deletes.

**Reply by stefanv** [reply]
You're right: when I add ACLs to a Role, I want the join table to provide the link between the two. I think what is confusing is that there are three potential places to specify propagation, and I cannot find an example in the docs (the association table & proxy examples seems subtly different).

**Reply by zzzeek** [reply]
OK so , the warning can just be disabled if you set that "overlaps" parameter.  Im not saying that's what you should do here, but the warning here is noting something that has already been in your code previously; there's a way to manipulate the relationships such that there's a conflicting value written into the row.   your code is probably not doing that, and if you are clean about how rows go in and out, just silence the warning if that's the most expedient thing.  the warning is just trying to encourage people to think about how they want data written and deleted from a row such as an association row here.

**Reply by stefanv** [reply]
Thanks very much for all the guidance. The warning served its purpose in making me concerned about what we are doing, but I think I can at least now write some developer guidance on how to use the three models without causing problems.

**Reply by zzzeek** [reply]
great, that's likely the first positive thing anyone's said about this one :)

**Reply by stefanv** [reply]
Better the devil you know!

**Reply by CaselIT** [reply]
@stefanv if you can think of a clearer wording for this warning feel free to open a PR to suggest the change.
It's raised here https://github.com/sqlalchemy/sqlalchemy/blob/b51a5f596299e1e52ca27b2db20394110f584257/lib/sqlalchemy/orm/relationships.py#L2758

**Reply by stefanv** [reply]
@CaselIT Perhaps it would be more helpful to document the "best practice" for this specific (perhaps common?) use-case? I.e., an association table accompanied by a model, to simplify adding single row associations, and to query across associations.

The [association object](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object) page is close, but then it has "Working with the association pattern in its direct form requires that child objects are associated with an association instance before being appended to the parent" which (for my use-case) defeats the purpose of simplicity, and is probably only necessary because of the extra field in that example?

I was surprised when I saw you can do this even (also in @zzzeek's answer above); my expectation was for the association object to merely act as a straightforward way to manipulate the association table.

Here are two example of how we use the association object (without any extra fields).

Query:

```python
        # Ensure user is not already a member of target group
        gu = (
            GroupUser.query.filter(GroupUser.group_id == group_id)
            .filter(GroupUser.user_id == user_id)
            .first()
        )
        if gu is not None:
            return self.error(f"User {user_id} is already a member of group {group_id}")
```

Single row association insertion:

```python
    # Add user to specified groups
    for group_id in group_ids:
        with session() as s:
            s.add(GroupUser(user_id=user.id, group_id=group_id))
```

My current feeling is that [we](https://github.com/skyportal/skyportal/) may be better off dropping the association object and simply using the association table directly, as described in the [SQLAlchemy docs: many-to-many relationships](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-many).

I may also need to replace myself with someone who actually understands SQLAlchemy :see_no_evil: :)

**Reply by zzzeek** [reply]
> The [association object](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#association-object) page is close, but then it has "Working with the association pattern in its direct form requires that child objects are associated with an association instance before being appended to the parent" 

that's a poorly worded paragraph as I didnt even read it correctly just now (and I wrote it).   what it means is that if the association object pattern is present, then you dont append "child" to "parent", you associate "child" with the association object and append *that* to the parent.

> which (for my use-case) defeats the purpose of simplicity, and is probably only necessary because of the extra field in that example?

well yes, the whole point of the association object pattern is that the association object table has additional fields.

if your roleacl table doesn't have any extra fields on it, then we normally expect people to use the plain many-to-many pattern where you don't deal at all with the "role_acl" table, it's handled transparently.  but it expects that you are appending Role / ACL objects together which typically involves those collections being loaded into memory first, unless you are using the "dynamic" relationship loader which tends to be important for many-to-many collection management. 


> 
> I was surprised when I saw you can do this even (also in @zzzeek's answer above); my expectation was for the association object to merely act as a straightforward way to manipulate the association table.

plain many-to-many is usually what people consider to be the most straightforward way, but at the same time I can see how your pattern of making a RoleACL object is straightforward too.  I would note that RoleACL(role_id=x, group_id=y) doesn't make use of any relationships so you could just not put any relationships on RoleACL directly if you only use role.acls and acl.roles.


> 
> Here are two example of how we use the association object (without any extra fields).
> 
> Query:
> 
> ```python
>         # Ensure user is not already a member of target group
>         gu = (
>             GroupUser.query.filter(GroupUser.group_id == group_id)
>             .filter(GroupUser.user_id == user_id)
>             .first()
>         )
>         if gu is not None:
>             return self.error(f"User {user_id} is already a member of group {group_id}")
> ```
> 
> Single row association insertion:
> 
> ```python
>     # Add user to specified groups
>     for group_id in group_ids:
>         with session() as s:
>             s.add(GroupUser(user_id=user.id, group_id=group_id))
> ```

you might also just use an insert() construct with a list of parameters:

```py
with session() as s:
    s.execute(
            group_user_table.insert(), 
           [{"user_id": user.id, "group_id": group_id} for group_id in group_ids]
    )
```



> 
> My current feeling is that [we](https://github.com/skyportal/skyportal/) may be better off dropping the association object and simply using the association table directly, as described in the [SQLAlchemy docs: many-to-many relationships](https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#many-to-many).

right, but then you wouldn't have that GroupUser insertion pattern that might be intuitive for your project.  either way.
